src/functions/routers/mod.ts

```diff
@@ -0,0 +1,175 @@
+    RequiredInput,
+    RequiredOutput
+  > {
+    this.routes.push([actionConstraint, handler]);
+    return this;
+  }
+
+  /**
+   * Returns a method handling routing of action payloads to the appropriate action handler.
+   * The output of export() should be attached to the `blockActions` export of your function.
+   */
+  export(): ActionHandler<typeof this.func.definition> {
+    return async (context) => {
+      const action: BlockAction = context.action;
+      const handler = this.matchHandler(action);
+      if (handler === null) {
+        // TODO: what do in this case?
+        // perhaps the user typo'ed the action id when registering their handler or defining their block.
+        console.warn(
+          `Received block action payload with action=${
+            JSON.stringify(action)
+          } but this app has no action handler defined to handle it!`,
+        );
+        return;
+      }
+      return await handler(context);
+    };
+  }
+
+  /**
+   * Return the first registered ActionHandler that matches the action ID string provided.
+   */
+  matchHandler(
+    action: BlockAction,
+  ): ActionHandler<typeof this.func.definition> | null {
+    for (let i = 0; i < this.routes.length; i++) {
+      const route = this.routes[i];
+      let [constraint, handler] = route;
+      // Handle different constraint types below
+      if (
+        constraint instanceof RegExp || constraint instanceof Array ||
+        typeof constraint === "string"
+      ) {
+        // Normalize simple string constraints to be an array of strings for consistency in handling inside this method.
+        constraint = normalizeConstraintToArray(constraint);
+        // Handle the case where the constraint is either a regex or an array of strings to match against action_id
+        if (matchBlockActionConstraintField(constraint, "action_id", action)) {
+          return handler;
+        }
+      } else {
+        // Assumes an object as a constraint (type BlockActionConstraintObject)
+        // Return first match *within* any of the defined fields on the constaint object, but ensure there is a match on *all* defined fields
+        // Effectively a logical AND across the action_id and block_id field(s)
+        let actionIDMatched = false;
+        let blockIDMatched = false;
+        // TODO: this whole next series of if blocks feels awkward; must be a better way...
+        if (constraint.action_id) {
+          actionIDMatched = matchBlockActionConstraintField(
+            normalizeConstraintToArray(constraint.action_id),
+            "action_id",
+            action,
+          );
+        }
+        if (constraint.block_id) {
+          blockIDMatched = matchBlockActionConstraintField(
+            normalizeConstraintToArray(constraint.block_id),
+            "block_id",
+            action,
+          );
+        }
+        if (
+          (constraint.block_id && constraint.action_id && blockIDMatched &&
+            actionIDMatched) ||
```

Comment:
As discussed, this check doesn't ensure that both `block_id` and `action_id` match since the below checks will pass if either matches.